### PR TITLE
Params classes

### PR DESCRIPTION
**Pull request by duncansmart**
For access to Controller Action/Page Handler parameter names

### CODE REVIEW

**Comment by MarkFl12** [comment]
I find myself in need of this feature, is there anything I can do to help get this merged and released?

File: src/R4Mvc.Tools/Services/ControllerGeneratorService.cs [hunk 1]
```diff
@@ -93,7 +93,8 @@ public ClassDeclarationSyntax GeneratePartialController(ControllerDefinition con
             var actionsExpression = controller.AreaKey != null
                 ? _settings.HelpersPrefix + "." + controller.AreaKey + "." + controller.Name
                 : _settings.HelpersPrefix + "." + controller.Name;
-            var controllerMethodNames = SyntaxNodeHelpers.GetPublicNonGeneratedControllerMethods(controller.Symbol).Select(m => m.Name).Distinct().ToArray();
+            var methods = SyntaxNodeHelpers.GetPublicNonGeneratedControllerMethods(controller.Symbol);
```
Comment (btihanyi) [inline src/R4Mvc.Tools/Services/ControllerGeneratorService.cs]: I would add a `.ToList()` call for performance reasons.

File: src/R4Mvc.Tools/Services/ControllerGeneratorService.cs [hunk 2]
```diff
@@ -121,7 +122,27 @@ public ClassDeclarationSyntax GeneratePartialController(ControllerDefinition con
...
+
+                 * [GeneratedCode, DebuggerNonUserCode]
+                 * public class ActionParamsClass_LogOn
+                 * {
+                 *     public readonly string returnUrl = "returnUrl";
+                 *     public readonly string model = "model";
+                 * }
+                */
+                .ForEach(methods.Where(m => m.Parameters.Any()), (c, m) => c
+                    .WithStaticFieldBackedProperty($"{m.Name}Params", $"ActionParamsClass_{m.Name}", SyntaxKind.PublicKeyword)
```
Comment (btihanyi) [inline src/R4Mvc.Tools/Services/ControllerGeneratorService.cs]: The 'Params' suffix should come from the settings json file with the name of 'ParamsPropertySuffix'.

File: src/R4Mvc.Tools/Services/ControllerGeneratorService.cs [hunk 3]
```diff
@@ -121,7 +122,27 @@ public ClassDeclarationSyntax GeneratePartialController(ControllerDefinition con
...
+                 * public ActionParamsClass_LogOn LogOnParams { get { return s_params_LogOn; } }
+
+                 * [GeneratedCode, DebuggerNonUserCode]
+                 * public class ActionParamsClass_LogOn
+                 * {
+                 *     public readonly string returnUrl = "returnUrl";
+                 *     public readonly string model = "model";
+                 * }
+                */
+                .ForEach(methods.Where(m => m.Parameters.Any()), (c, m) => c
```
Comment (btihanyi) [inline src/R4Mvc.Tools/Services/ControllerGeneratorService.cs]: Parameter generation should be optional (as in the original T4MVC), which would be decided from the settings json file with the name 'GenerateParamsForActionMethods'.

**Review by btihanyi — CHANGES_REQUESTED** [review]
Nice work, I also need this feature for upgrading to ASP.NET Core.

**Comment by artiomchi** [comment]
Hey all,
Thanks a lot for your contribution @duncansmart, and the review on this @btihanyi 
I've applied the comments and am merging this PR now.

Since #148 was recently merged, I'm applying those changes too, so if the param name was renamed, the params class will have the updated value (but the name will match the param name in code)

**Pull request merged**